gobblin-modules/gobblin-crypto/src/main/java/org/apache/gobblin/crypto/GPGCodec.java

```diff
@@ -20,31 +20,74 @@
 import java.io.InputStream;
 import java.io.OutputStream;
 
-import org.apache.gobblin.codec.StreamCodec;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
 
+import org.apache.gobblin.codec.StreamCodec;
 
 /**
- * Codec class that supports GPG decryption (only).
+ * Codec class that supports GPG encryption and decryption.
  */
 public class GPGCodec implements StreamCodec {
   public static final String TAG = "gpg";
 
   private final String password;
+  private final String cipher;
+  private final Path keyRingPath;
+  private final long keyId;
+  private final Configuration conf;
 
-  public GPGCodec(String password) {
+  /**
+   * Constructor for a {@code GPGCodec} configured for password-based encryption
+   * @param password password for symmetric encryption
+   * @param cipher the symmetric cipher to use for encryption. If null or empty then a default cipher is used.
+   */
+  public GPGCodec(String password, String cipher) {
     this.password = password;
+    this.cipher = cipher;
+
+    this.keyRingPath = null;
+    this.keyId = 0;
+    this.conf = null;
+  }
+
+  /**
+   * Constructor for a {@code GPGCodec} configured for key-based encryption
+   * @param keyRingPath path to the key ring
+   * @param passphrase passphrase for retrieving the decryption key
+   * @param keyId id for retrieving the key used for encryption
+   * @param cipher the symmetric cipher to use for encryption. If null or empty then a default cipher is used.
+   */
+  public GPGCodec(Path keyRingPath, String passphrase, long keyId, String cipher) {
+    this.keyRingPath = keyRingPath;
+    this.password = passphrase;
+    this.keyId = keyId;
+    this.cipher = cipher;
+    this.conf = new Configuration();
   }
 
   @Override
   public OutputStream encodeOutputStream(OutputStream origStream)
       throws IOException {
-    throw new UnsupportedOperationException("Not implemented yet");
+    if (this.keyRingPath != null) {
+      try (InputStream keyRingInputStream = this.keyRingPath.getFileSystem(this.conf).open(this.keyRingPath)) {
+        return GPGFileEncryptor.encryptFile(origStream, keyRingInputStream, this.keyId, this.cipher);
```

Comment:
It looks a little strange to pass `keyRingInputStream` into `GPGFileEncryptor` and then close it even if no exception is thrown -- should we be defensive and assume that the `GPGFileEncryptor` takes ownership of the stream if `encryptFile` succeeds (and therefore not close it ourselves)?

Reply:
The key ring is read to construct a `PGPPublicKeyRingCollection`. The APIs that do this don't close the input stream. I think it makes sense for the close to be coupled with the open here since GPGFileEncryptor is a utility class and GPGFileEncryptor.encryptFile() does not keep any references to the InputStream.

Reply:
Added comment indicating that the caller is the owner for the keyring inputstream.